src/System.Net.Security/src/System/Net/Security/SslState.cs

```diff
@@ -1346,25 +1358,34 @@ internal void FinishWrite()
 
             lock (this)
             {
-                object obj = _queuedWriteStateRequest;
-                if (obj == null)
-                {
-                    // A repeated call.
-                    return;
-                }
+                HandleQueuedCallback(ref _queuedWriteStateRequest);
+            }
+        }
 
-                _queuedWriteStateRequest = null;
-                if (obj is LazyAsyncResult)
-                {
-                    // Sync handshake is waiting on other thread.
-                    ((LazyAsyncResult)obj).InvokeCallback();
-                }
-                else
-                {
-                    // Async handshake is pending, start it on other thread.
-                    // Consider: we could start it in on this thread but that will delay THIS write completion
+        private void HandleQueuedCallback(ref object queuedStateRequest)
+        {
+            if(queuedStateRequest == null)
+            {
+                return;
+            }
+                   
+            object obj = queuedStateRequest;
```

Comment:
How about moving this up to the beginning and then having the if condition check `obj` instead of `queuedStateRequest`?